src/util/vecdeque.h

```diff
@@ -0,0 +1,313 @@
+ *   unless reallocation is needed (in which case they are O(n)).
+ * - Supports reserve(), capacity(), shrink_to_fit() like vectors.
+ * - No iterator support.
+ * - Data is not stored in a single contiguous block, so no data().
+ */
+template<typename T>
+class VecDeque
+{
+    /** Pointer to allocated memory. Can contain constructed and uninitialized T objects. */
+    T* m_buffer{nullptr};
+    /** m_buffer + m_offset points to first object in queue. m_offset = 0 if m_capacity is 0;
+     *  otherwise 0 <= m_offset < m_capacity. */
+    size_t m_offset{0};
+    /** Number of objects in the container. 0 <= m_size <= m_capacity. */
+    size_t m_size{0};
+    /** The size of m_buffer, expressed as a multiple of the size of T. */
+    size_t m_capacity{0};
+
+    /** Returns the number of populated objects between m_offset and the end of the buffer. */
+    size_t FirstPart() const noexcept { return std::min(m_capacity - m_offset, m_size); }
+
+    void Reallocate(size_t capacity)
+    {
+        Assume(capacity >= m_size);
+        Assume((m_offset == 0 && m_capacity == 0) || m_offset < m_capacity);
+        // Allocate new buffer.
+        T* new_buffer = capacity ? std::allocator<T>().allocate(capacity) : nullptr;
+        if (capacity) {
+            if constexpr (std::is_trivially_copyable_v<T>) {
+                // When T is trivially copyable, just copy the data over from old to new buffer.
+                size_t first_part = FirstPart();
+                if (first_part != 0) {
+                    std::memcpy(new_buffer, m_buffer + m_offset, first_part * sizeof(T));
+                }
+                if (first_part != m_size) {
+                    std::memcpy(new_buffer + first_part, m_buffer, (m_size - first_part) * sizeof(T));
+                }
+            } else {
+                // Otherwise move-construct in place in the new buffer, and destroy old buffer objects.
+                size_t old_pos = m_offset;
+                for (size_t new_pos = 0; new_pos < m_size; ++new_pos) {
+                    std::construct_at(new_buffer + new_pos, std::move(*(m_buffer + old_pos)));
+                    std::destroy_at(m_buffer + old_pos);
+                    ++old_pos;
+                    if (old_pos == m_capacity) old_pos = 0;
+                }
+            }
+        }
+        // Deallocate old buffer and update housekeeping.
+        std::allocator<T>().deallocate(m_buffer, m_capacity);
+        m_buffer = new_buffer;
+        m_offset = 0;
+        m_capacity = capacity;
+        Assume((m_offset == 0 && m_capacity == 0) || m_offset < m_capacity);
+    }
+
+    /** What index in the buffer does logical entry number pos have? */
+    size_t BufferIndex(size_t pos) const noexcept
+    {
+        if (pos >= m_capacity - m_offset) {
+            return pos - (m_capacity - m_offset);
+        } else {
+            return pos + m_offset;
+        }
```

Comment:
For me this is more straight forward:
```suggestion
        if (m_offset + pos >= m_capacity) {
            return (m_offset + pos) - m_capacity;
        } else {
            return m_offset + pos;
        }
```
"We take where we begin now, `m_offset`, add `pos` to it and see what that means."
Subtractions hurt readability in this case.

Reply:
Excellent catch! I got a bit afraid that `(m_offset + pos) - m_capacity` might also run into issues when overflowing but appears to work out (and unsigned integer overflow is not undefined behavior).